### PR TITLE
[Reviewer: RKD] Docker split-storage

### PR DESCRIPTION
**Pull request by MatMeredith**
This change is [Reviewable](https://reviewable.io/reviews/metaswitch/clearwater-docker/29)

### CODE REVIEW

**Comment by MatMeredith** [comment]
- Split out storage services
- Add memcached_user_plugin on Sprout and Ralf.   This watches the memcached cluster config in etcd and updates the cluster_settings so that Sprout and Ralf pick up on changes to the memcached cluster.
- Add the ability to specify an external IP address for Bono
- Make cassandra start up consistent with other services -- use /etc/init.d/cassandra rather than running cassandra directly
- Remove stop / start sections from readme -- stopping and starting the containers changes IP addresses and breaks everything
- Remove links from compose file -- these aren't necessary and break load balancing when you have more than one of a given node type.
- Add a few utility scripts to help with monitoring the cluster

File: sprout/Dockerfile [hunk 1]
```diff
@@ -4,9 +4,9 @@ MAINTAINER maintainers@projectclearwater.org
 RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --force-yes sprout
 RUN sed -e 's/\(echo 0 > \/proc\/sys\/kernel\/yama\/ptrace_scope\)/# \0/g' -i /etc/init.d/sprout
 
-COPY chronos.supervisord.conf /etc/supervisor/conf.d/chronos.conf
-COPY memcached.supervisord.conf /etc/supervisor/conf.d/memcached.conf
 COPY sprout.supervisord.conf /etc/supervisor/conf.d/sprout.conf
 COPY clearwater-group.supervisord.conf /etc/supervisor/conf.d/clearwater-group.conf
+COPY plugins/* /usr/share/clearwater/clearwater-cluster-manager/plugins/
```
Comment (rkday-pro) [inline sprout/Dockerfile:9]: Why is this necessary? Shouldn't other packages provide this?

Comment (MatMeredith) [reply sprout/Dockerfile:9]: @rkd-msw No -- this is my new memcached-user-plugin.   It's only needed in Docker and so I've defined it here in the clearwater-docker repo?

**Comment by MatMeredith** [comment]
@rkd-msw I've changed the way Cassandra starts up (again).   One of the reasons we were seeing failures previously, is that if you start Cassandra in the foreground then it doesn't write a pidfile, and so the cassandra_plugin can't kill Cassandra before deleting all of it's data (per the stack in Slack the other day that you commented on).   Now I've gone back to starting Cassandra in the background, but controlling it from a blocking shell script.   One question: I wasn't sure where in the container to put the script so I went for /usr/bin/.  I don't know if that's the correct place though?

Can you re-review please?   I've also backed out my changes to clearwater-cassandra to add support for "run".

File: cassandra/start_cassandra.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    kill $(cat $pidfile)
+    exit 0 # exit okay
+}
+trap "kill_app" SIGINT SIGTERM
+
+# Launch daemon
+$command
+sleep 2
+
+# Loop while the pidfile and the process exist
```
Comment (rkday-pro) [inline cassandra/start_cassandra.sh:18]: `wait $(cat pidfile)`?

File: .env [hunk 3]
```diff
@@ -0,0 +1 @@
+PUBLIC_IP=
```
Comment (rkday-pro) [inline .env:1]: Is the .env filename special in any way? It feels odd to have this as a hidden file.

File: README.md [hunk 4]
```diff
@@ -20,6 +20,8 @@ To prepare your system to deploy Clearwater on Docker, run:
 
     # Checkout clearwater-docker.
     git clone --recursive git@github.com:Metaswitch/clearwater-docker.git
+    
+Edit clearwater-docker/.env so that PUBLIC_IP is set to an IP address that can be used by SIP clients to access the docker host.   E.g. if you are running in AWS, this wants to be the public IP of your AWS VM.
```
Comment (rkday-pro) [inline README.md:24]: Maybe worth noting that this appears in Record-Route headers and calls will drop after 30s if this is wrong?

File: cassandra/start_cassandra.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#! /usr/bin/env bash
```
Comment (rkday-pro) [inline cassandra/start_cassandra.sh:1]: I wonder whether this should actually live in clearwater-cassandra (and be caled start_cassandra_foregrounded), but the /usr/bin location is fine.

File: ralf/plugins/memcached_user_plugin.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# Project Clearwater - IMS in the Cloud
+# Copyright (C) 2015  Metaswitch Networks Ltd
```
Comment (rkday-pro) [inline ralf/plugins/memcached_user_plugin.py:2]: You've checked in the corresponding .pyc files - worth deeting and .gitignore'ing them.

File: sprout/Dockerfile (hunk 1 above)
Comment (rkday-pro) [reply sprout/Dockerfile:9]: Arguably this isn't only needed in Docker - it's needed anywhere we split memcached out but don't use Astaire as a proxy - and so it might be worth putting these in a separate package (so they can be deployed in e.g. OpenStack if we want to demo this topology there). As we're merging the Astaire stuff in a few weeks anyway, that's probably not worth doing.

File: utils/short_ps.sh [hunk 7]
```diff
@@ -0,0 +1 @@
+sudo docker ps --format "table {{.Names}}\t{{.Image}}\t{{.Ports}}"
```
Comment (rkday-pro) [inline utils/short_ps.sh:1]: We wouldn't normally have sudo here, we'd expect the user to run this with sudo - that makes it more explicit that this has eevated permissions.

**Review by rkday-pro — APPROVED** [review]
Looks good.
